hw/ip/otbn/util/shared/information_flow_analysis.py

```diff
@@ -0,0 +1,598 @@
+        edges = [ControlLoc(section.end + 4)]
+    else:
+        # Update the constants to include the last instruction
+        _update_constants_insn(last_insn, last_op_vals, constants)
+
+    # We're only returning constants that are the same in all RET branches
+    common_constants = None
+
+    for loc in edges:
+        if isinstance(loc, Ecall) or isinstance(loc, ImemEnd):
+            if len(edges) != 1:
+                raise RuntimeError(
+                    'Control graph section at PC {:#x} has edges {}; if '
+                    'edges contain an Ecall or ImemEnd, it is expected to '
+                    'be the only edge.'.format(section.end, edges))
+
+            # Clear common constants; there are no return branches here
+            common_constants = {'x0': 0}
+            if program_end_iflow is None:
+                program_end_iflow = iflow
+            else:
+                program_end_iflow.update(iflow)
+        elif isinstance(loc, Ret):
+            if loop_end_pc is not None:
+                raise RuntimeError(
+                    'RET before end of loop at PC {:#x} (loop end PC: '
+                    '{:#x})'.format(section.end, loop_end_pc))
+            if len(edges) != 1:
+                raise RuntimeError(
+                    'Control graph section at PC {:#x} has edges {}; if '
+                    'edges contain a Ret, it is expected to be the only '
+                    'edge.'.format(section.end, edges))
+
+            # Since this is the only edge, common_constants must be unset
+            common_constants = constants
+            if return_iflow is None:
+                return_iflow = iflow
+            else:
+                return_iflow.update(iflow)
+
+        elif isinstance(loc, LoopStart):
+            # We shouldn't hit a LoopStart here; those cases (a loop
+            # instruction or the end of a loop) are all handled earlier
+            raise RuntimeError('Unexpected LoopStart at PC {:#x}'.format(
+                section.end))
+        elif not loc.is_special():
+            # Just a normal PC; recurse
+            result = _get_iflow(program, graph, loc.pc, constants, loop_end_pc,
+                                cache)
+
+            # Update program_end_iflow, used constants, and control_deps
+            program_end_iflow = _get_iflow_update_state(
+                result, iflow, program_end_iflow, used_constants, control_deps)
+
+            # Get information flow for return paths and new constants
+            _, rec_return_iflow, _, rec_constants, _ = result
+
+            # Update common constants
+            if common_constants is None:
+                common_constants = rec_constants
+            else:
+                for const, value in rec_constants.items():
+                    if common_constants.get(const, None) != value:
+                        # Remove from common_constants if key exists
+                        common_constants.pop(const, None)
```

Comment:
This sort of merge operation really looks like it would be nice to wrap up in a "constants" datatype. Not necessarily in this PR, but it's probably worth thinking about. You could also get rid of the Optional bit here because the "I've got no data" value would just be represented as some possible value of the datatype's fields.

Reply:
I'm not sure I understand the last comment about the Optional bit, could you elaborate? Do you mean the "constants" datatype would also have a "single constant" datatype that can be "no data"?

Reply:
I just mean that at the moment `common_constants` might be `None`. I'm suggesting that if you wrap it up in a nice datatype, you could have uniform support for the value that you're currently representing as `None` as well as everything else.

Definitely doesn't need to be in this PR, but I think that you could probably shove quite a bit of this complexity into update methods on data types representing the various bits of data that you're plumbing through. The result should hopefully be that the "application logic" in this function gets much more obvious to follow.

Reply:
Ah, I think I see what you mean now. And yes, I think you're right that some of these bits of state are fiddly enough that it's worth packaging their management into purpose-build datatypes with different update functions, so that complexity is less exposed (and the main `_get_iflow` function is less ridiculously long)!